kernels/mpi/src/observables/MPIObservables.cpp

```diff
@@ -33,216 +33,356 @@
 
 /**
- * « detailed description »
+ * Implementation of observables for the MPI kernel. In most cases during the evaluate(),
+ * the workers collect the results which are then gathered on the master worker.
  *
  * @file MPIObservables.cpp
- * @brief « brief description »
+ * @brief Implementation of observables for the MPI kernel
  * @author chrisfroe
  * @date 03.06.19
  */
 
+#include <utility>
 #include <readdy/kernel/mpi/observables/MPIObservables.h>
 #include <readdy/kernel/mpi/MPIKernel.h>
 
-#include <utility>
-
 namespace readdy::kernel::mpi::observables {
 
-MPIVirial::MPIVirial(MPIKernel *kernel, Stride stride) : Virial(kernel, stride), kernel(kernel) {
-}
+MPIVirial::MPIVirial(MPIKernel *kernel, Stride stride) : Virial(kernel, stride), kernel(kernel) {}
 
 void MPIVirial::evaluate() {
-    // @todo MPI gather results, only save on master rank
-    result = kernel->getMPIKernelStateModel().virial();
+    // todo use MPI reduce with the standard '+' op, MPI_SUM
+    //MPI_Reduce(send_data, recv_data, count, datatype, op, root, communicator);
+    if (not kernel->domain().isIdleRank()) {
+        std::vector<readdy::model::observables::Virial::result_type> results(1);
+        if (kernel->domain().isWorkerRank()) {
+            results[0] = kernel->getMPIKernelStateModel().virial();
+        }
+        results = util::gatherObjects(results, 0, kernel->domain(), kernel->commUsedRanks());
+
+        if (kernel->domain().isMasterRank()) {
+            // add up virial tensors assuming that there was no double counting
+            // which has to be ensured in calculateForces
+            result = std::accumulate(results.begin(), results.end(), _internal::ReaDDyMatrix33<scalar>());
+        }
+    }
+}
+
+void MPIVirial::append() {
+    if (kernel->domain().isMasterRank()) {
+        Virial::append();
+    }
+}
+
+void MPIVirial::initializeDataSet(File &file, const std::string &dataSetName, Stride flushStride) {
+    if (kernel->domain().isMasterRank()) {
+        Virial::initializeDataSet(file, dataSetName, flushStride);
+    }
 }
 
 MPIPositions::MPIPositions(MPIKernel *kernel, unsigned int stride, const std::vector<std::string> &typesToCount)
         : Positions(kernel, stride, typesToCount), kernel(kernel) {}
 
 void MPIPositions::evaluate() {
-    // @todo MPI gather results, only save on master rank
-    result.clear();
-    auto &stateModel = kernel->getMPIKernelStateModel();
-    const auto &pd = stateModel.getParticleData();
-    auto it = pd->cbegin();
-    if (typesToCount.empty()) {
-        result = stateModel.getParticlePositions();
-    } else {
-        // only get positions of typesToCount
-        while (it != pd->cend()) {
-            if (!it->is_deactivated()) {
-                if (std::find(typesToCount.begin(), typesToCount.end(), it->type) != typesToCount.end()) {
-                    result.push_back(it->position());
+    if (not kernel->domain().isIdleRank()) {
```

Comment:
perhaps it would be good to filter the idle ranks out further up in the algorithm chain so they dont even appear here

Reply:
yes sounds reasonable